### ISSUE TITLE
`F704` wrong "await outside of a function" when using generators

### ISSUE THREAD

**Issue by PabloRuizCuevas** [astral-sh/ruff#21750]
### Describe the Bug
Await can be outside of a function if is in code that won't be executed automatically (ex: async generators).

> `await` statement outside of a function

This code is correct, yet ruff complains:

```python
import asyncio

async def test2():
    return "test"

async def a_gen():
    for i in range(3):
        yield test2()

a = (await cor async for cor in a_gen()) # `await` statement outside of a function F704

async def main():
    async for i in a:
        print(i)

if __name__ == "__main__":
    asyncio.run(main())
```

I suggest changing F704  so awaits are ok inside async generators that are basically syntax sugar of a function.

of course for now we can do the equivalent:

```python
async def a():
    for cor in a_gen():
        yield await cor
```

**Comment by MichaReiser** [comment]
Thanks. I consider this a bug in our semantic rules. I'm a bit surprised that `F704` flags awaits, given that we also have `PLE1142` (which correctly detects that we shouldn't flag this await usage).

**Comment by ntBre** [comment]
I think we'll just need to update our `in_await_allowed_context` methods to return true for async generators:

https://github.com/astral-sh/ruff/blob/0e651b50b7d0c2614c5958e78525213e759e5b41/crates/ruff_linter/src/checkers/ast/mod.rs#L777-L782

It looks like I just missed this case in #17298.

**Comment by MichaReiser** [comment]
Might be worth a bug fix PR if that's indeed all that's necessary
